plugins/storage/volume/storpool/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/StorpoolDownloadTemplateCommandWrapper.java

```diff
@@ -0,0 +1,133 @@
+            }
+
+            final NfsTO nfsImageStore = (NfsTO)srcDataStore;
+            final String tmplturl = nfsImageStore.getUrl() + File.separator + src.getPath();
+            final int index = tmplturl.lastIndexOf("/");
+            final String mountpoint = tmplturl.substring(0, index);
+            String tmpltname = null;
+            if (index < tmplturl.length() - 1) {
+                tmpltname = tmplturl.substring(index + 1);
+            }
+
+            secondaryPool = storagePoolMgr.getStoragePoolByURI(mountpoint);
+
+            KVMPhysicalDisk srcDisk = null;
+
+            if (tmpltname == null) {
+                secondaryPool.refresh();
+                final List<KVMPhysicalDisk> disks = secondaryPool.listPhysicalDisks();
+                if (disks == null || disks.isEmpty()) {
+                    SP_LOG("Failed to get volumes from pool: " + secondaryPool.getUuid());
+                    return new CopyCmdAnswer("Failed to get volumes from pool: " + secondaryPool.getUuid());
+                }
+                for (final KVMPhysicalDisk disk : disks) {
+                    if (disk.getName().endsWith("qcow2")) {
+                        srcDisk = disk;
+                        break;
+                    }
+                }
+            } else {
+                srcDisk = secondaryPool.getPhysicalDisk(tmpltname);
+            }
+
+            if (srcDisk == null) {
+                SP_LOG("Failed to get template from pool: " + secondaryPool.getUuid());
+                return new CopyCmdAnswer("Failed to get template from pool: " + secondaryPool.getUuid());
+            }
+
+            SP_LOG("got src path: " + srcDisk.getPath() + " srcSize " + srcDisk.getVirtualSize());
+
+            final QemuImgFile srcFile = new QemuImgFile(srcDisk.getPath(), srcDisk.getFormat());
+
+            final QemuImg qemu = new QemuImg(cmd.getWaitInMillSeconds());
+            StorpoolStorageAdaptor.resize( Long.toString(srcDisk.getVirtualSize()), dst.getPath());
+
+            dstPath = dst.getPath();
+            StorpoolStorageAdaptor.attachOrDetachVolume("attach", cmd.getObjectType(), dstPath);
+
+            final QemuImgFile dstFile = new QemuImgFile(dstPath, PhysicalDiskFormat.RAW);
+
+            qemu.convert(srcFile, dstFile);
+            return new CopyCmdAnswer(dst);
+        } catch (final Exception e) {
+            final String error = "Failed to copy template to primary: " + e.getMessage();
+            s_logger.debug(error);
+            return new CopyCmdAnswer(error);
+        } finally {
+            if (dstPath != null) {
+                StorpoolStorageAdaptor.attachOrDetachVolume("detach", cmd.getObjectType(), dstPath);
+            }
+
+            if (secondaryPool != null) {
+                try {
+                    secondaryPool.delete();
+                } catch (final Exception e) {
+                    s_logger.debug("Failed to delete secondary storage", e);
+                }
+            }
```

Comment:
This logic is repeated along several classes, but changing some values, therefore, it could be a method in a helper or util.